HPCC-16451 Fix Coverity warnings in hidl generated ESP code

afishbeck opened the pull request:
Warnings related to dynamic_cast pointers and uninitialized member
variables.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck committed:
HPCC-16451 Fix Coverity warnings in hidl generated ESP code

Warnings related to dynamic_cast pointers and uninitialized member
variables.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest
Sha: 5d732fd232a305a90ba19a446533403c26c1542d
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.136 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@ghalliday Please review

ghalliday commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -4268,7 +4268,11 @@ void EspMessageInfo::write_esp()
     //method ==> copy
     outf("\nvoid C%s::copy(C%s &from)\n{\n", name_, name_);
     if (parent)
-        outf("\tC%s::copy(*dynamic_cast<C%s*>(&from));\n", parent, parent);
```
minor:  I think this can be a static cast since C<name> will be derived from C<parent> if parent is non null.

afishbeck replied:
@ghalliday I think these gave me compiler errors when I tried to make them static_cast.  But I can investigate further.

ghalliday commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -4426,7 +4434,9 @@ void EspMessageInfo::write_esp()
     if (parent)
     {
-        outf("\tC%s::serializer(ctx,dynamic_cast<C%s&>(src),buffer,false);\n",parent,parent);
+        outf(1, "C%s *baseSrc = dynamic_cast<C%s*>(&src);\n", parent, parent);
```
similarly this can be an unconditional static cast

ghalliday reviewed (commented):
Generally look like good changes.  A couple of items to double check whether they can be static casts instead.

ghalliday commented:
@afishbeck a couple of comments

afishbeck committed:
HPCC-16451 Further improvements after review

Use two step static_cast thunking rather than dynamic_cast where possible.

I think two static_casts is still a huge performance gain over
dynamic_cast.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -4426,7 +4433,9 @@ void EspMessageInfo::write_esp()
     {
-        outf("\tC%s::serializer(ctx,dynamic_cast<C%s&>(src),buffer,false);\n",parent,parent);
+        outf(1, "C%s *classSrc = static_cast<C%s*>(&src);\n", name_, name_);
+        outf(1, "C%s *baseSrc = static_cast<C%s*>(classSrc);\n", parent, parent);
```
@ghalliday same here.

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -4278,7 +4281,11 @@ void EspMessageInfo::write_esp()
-        outf("\tC%s::copy(*dynamic_cast<IConst%s*>(&ifrom));\n", parent, parent);
+    {
+        outf(1, "C%s *classFrom = static_cast<C%s*>(&ifrom);\n", name_, name_);
+        outf(1, "IConst%s *baseICFrom = static_cast<IConst%s*>(classFrom);\n", parent, parent);
```
@ghalliday a single static_cast wouldn't work but I think two static_casts is still much faster that dynamic_cast.

ghalliday replied:
yes, it will be much faster - it should be optimized to be equivalent to a single cast.

HPCCSmoketest commented:
Automated Smoketest
Sha: 1fa52af8c7709a72b6c2c9c209acf78194ee217f
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.306 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[31m [ FAILED  ]  \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[32m [   OK    ]  \x1b[0m 
HPCC status:
mydafilesrv     ( pid     5282 ) is running ...
mydali          ( pid     5435 ) is running ...
mydfuserver     ( pid     5605 ) is running ...
myeclagent      ( pid     5778 ) is running ...
myeclccserver   ( pid     5939 ) is running ...
myeclscheduler  ( pid     6107 ) is running ...
myesp           is stopped,
myroxie         ( pid     6447 ) is running ...
mysasha         ( pid     6634 ) is running ...
mythor          ( pid     7053 ) is running with 1 slave process(es) ...
myesp           is stopped,
pushd /mnt/disk1/home/vamosax/smoketest/PR-9407
~/smoketest/PR-9407 ~/smoketest/PR-9407/build
./unittest.sh
~/smoketest/PR-9407/build
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
Looks good.  I'll merge and then rerun the coverity analysis.
